File: demos_code_J2_conception_objet/01_dip_port_adapter/DemoDipAvant.java
import java.math.BigDecimal;

public class DemoDipAvant {
    static final class StripeSdk {
        String charge(int cents, String token) {
            return "stripe_tx_" + cents + "_" + token.substring(0, 4);
        }
    }

    static final class CheckoutService {
        // Le cas d'usage dépend DIRECTEMENT d'un détail fournisseur.
        private final StripeSdk stripe = new StripeSdk();

        String checkout(BigDecimal amount, String cardToken) {
            int cents = amount.movePointRight(2).intValueExact();
            String stripeId = stripe.charge(cents, cardToken);
            return "Paiement accepté : " + stripeId;
        }
    }

    public static void main(String[] args) {
        System.out.println(new CheckoutService().checkout(new BigDecimal("42.50"), "tok_demo_123"));
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/01_dip_port_adapter/demo_dip_avant.py
from decimal import Decimal


class StripeSdk:
    def charge(self, cents: int, token: str) -> str:
        return f"stripe_tx_{cents}_{token[:4]}"


class CheckoutService:
    def __init__(self) -> None:
        # Le cas d'usage dépend DIRECTEMENT d'un détail fournisseur.
        self._stripe = StripeSdk()

    def checkout(self, amount: Decimal, card_token: str) -> str:
        cents = int(amount * 100)
        stripe_id = self._stripe.charge(cents, card_token)
        return f"Paiement accepté : {stripe_id}"


if __name__ == "__main__":
    print(CheckoutService().checkout(Decimal("42.50"), "tok_demo_123"))
